horovod/torch/mpi_ops.cc

```diff
@@ -0,0 +1,340 @@
+// Copyright 2018 Uber Technologies, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <chrono>
+#include <memory>
+#include <thread>
+
+#include "../common/operations.h"
+#include "adapter.h"
+#include "handle_manager.h"
+#include "mpi_ops.h"
+#include "ready_event.h"
+#include "cuda_util.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace torch {
+
+static HandleManager handle_manager;
+
+template <class T>
+int DoAllreduce(T* tensor, T* output, int average, char* name) {
+  ThrowIfError(common::CheckInitialized());
+
+  auto handle = handle_manager.AllocateHandle();
+  auto name_or_handle =
+      name != nullptr ? std::string(name) : "noname." + std::to_string(handle);
+
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_tensor = std::make_shared<TorchTensor<T>>(tensor);
+  auto hvd_context = std::make_shared<TorchOpContext<T>>(device, output);
+  auto hvd_output = std::make_shared<TorchTensor<T>>(output);
+
+  auto enqueue_result = EnqueueTensorAllreduce(
+      hvd_context, hvd_tensor, hvd_output, nullptr,
+      "allreduce." + name_or_handle, device,
+      [handle, average, output](const Status& status) {
+        if (average) {
+          TensorUtil::DivideTensorInPlace(output, horovod_size());
+        }
+        handle_manager.MarkDone(handle, status);
+      });
+  ThrowIfError(enqueue_result);
+
+  return handle;
+}
+
+#if HAVE_CUDA
+template <class TC, class T>
+int DoAllreduceCudaOnCPU(TC* tensor, TC* output, int average, char* name) {
+  ThrowIfError(common::CheckInitialized());
+
+  auto handle = handle_manager.AllocateHandle();
+  auto name_or_handle =
+      name != nullptr ? std::string(name) : "noname." + std::to_string(handle);
+
+  // Make async copy of input tensor to CPU tensor and record completion event.
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_cpu_buffer =
+      std::make_shared<TorchTemporaryBuffer<T>>(CPU_DEVICE_ID);
+  TensorUtil::AsyncCopyCudaToCPU(tensor, hvd_cpu_buffer->tensor());
+  auto ready_event = std::make_shared<TorchReadyEvent<TC>>(device);
+
+  auto hvd_context = std::make_shared<TorchOpContext<T>>(
+      CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
+
+  auto enqueue_result = EnqueueTensorAllreduce(
+      hvd_context, hvd_cpu_buffer, hvd_cpu_buffer, ready_event,
+      "allreduce." + name_or_handle, CPU_DEVICE_ID,
```

Comment:
Can you write a helper function to return `"allreduce." + (name != nullptr ? std::string(name) : "noname." + std::to_string(handle))`